contracts/core/OriginCore.sol

```diff
@@ -414,4 +481,135 @@ contract OriginCore is OriginCoreInterface, OriginCoreConfig {
+                prefix,
+                _voteHash
+            )
+        );
+
+        signer_ = ecrecover(
+            _voteHash,
+            _v,
+            _r,
+            _s
+        );
+
+        /* Header of last meta block. */
+        MetaBlock.Header storage latestMetaBlockHeader = reportedHeaders[head];
+        /* Current meta-block height. */
+        currentHeight_ = latestMetaBlockHeader.kernel.height.add(1);
+        uint256 weight = stake.weight(currentHeight_, signer_);
+
+        require(
+            weight > 0,
+            "Only validator with non zero weight can vote."
+        );
+
+        require(
+            seals[_transitionHash].validators[signer_] == false,
+            "Vote already verified for this validator."
+        );
+
+    }
+
+    /**
+     * @notice This function saves vote into the seal.
+     *
+     * @dev This method assumes that vote has been successfully verified. It
+     *      will revert the transaction if validator already voted for this
+     *      transition object.
+     *
+     * @param _transitionHash The hash of the transition part of the meta-block
+     *                    header at the source block.
+     * @param _signer Address of validator who has signed vote object.
+     * @param _height Height of current meta-block.
+     */
+    function saveVote(
+        bytes32 _transitionHash,
+        address _signer,
+        uint256 _height
+    )
+        private
+        returns(uint256 totalVoteWeight_)
+    {
+        MetaBlock.Seal storage seal = seals[_transitionHash];
+
+        seal.validators[_signer] = true;
+
+        totalVoteWeight_ = seal.totalVoteWeight.add(stake.weight(_height,_signer));
+
+        seal.totalVoteWeight = totalVoteWeight_;
+
+        //seals[_transition] = seal;
+    }
+
+    /**
+     * @notice This function checks if meta-block proposal exists.
+     *
+     * @param _kernelHash The hash of the current kernel.
+     * @param _transitionHash The hash of the transition object which is proposed
+     *                    with meta-block.
+     * @return bool `true` If meta-block proposal exists.
+     */
+    function existsMetaBlockProposal(
+        bytes32 _kernelHash,
+        bytes32 _transitionHash
+    )
+        private
+        view
+        returns(bool)
+    {
+        MetaBlock.AuxiliaryTransition storage transitionObject =
+            proposedMetaBlock[_kernelHash][_transitionHash];
+
+        return transitionObject.kernelHash == _kernelHash;
+    }
+
+    /**
+     * @notice Function to calculated weight required for super majority
+     *         i.e 2/3rd of total weight.
+     *
+     * @param _totalWeight Total weight of all the validators at current
+     *                     meta-block height.
+     * @return uint256 Required weight for 2/3rd super majority.
+     */
+    function requiredWeight(uint256 _totalWeight)
+        private
+        pure
+        returns(uint256)
+    {
+        return _totalWeight.mul(2).div(3);
```

Comment:
See `PollingPlace::requiredWeight()`. There is an edge-case as integer division in Solidity always rounds down. Possibly extract into shared library.